### PR TITLE
Diesel query fragment type inference

### PR DESCRIPTION
**Pull request by Ten0**
As discussed in https://github.com/diesel-rs/diesel/discussions/3661

This introduces:
- The `#[dsl::auto_type]` attribute macro, allowing to infer type of query fragment functions
- Using the same type inference on `Selectable` derives, which allows skipping specifying `select_expression_type` most of the time, in turn enabling most queries to be written using just a `Selectable` derive.

### CODE REVIEW

**Comment by Ten0** [comment]
So I've tested this on our codebase and it feels amazing :)
That makes it really easy to write most queries using `Selectable` derives.

File: diesel_derives/src/lib.rs
Comment (weiznich) [inline diesel_derives/src/lib.rs:28]: I would prefer to stay with non-nested imports here, just to be consistent with the rest of the code base.

File: diesel_derives/src/lib.rs
Comment (weiznich) [inline diesel_derives/src/lib.rs:886]: Maybe list cases that continue to require this? (I would assume that if `select_expression` is a custom function call it might be required?)

File: diesel_derives/src/lib.rs
Comment (weiznich) [inline diesel_derives/src/lib.rs:1696]: We might want to be explicit here that this is "required" if you mix your own custom query dsl extensions with diesel types. Maybe we should just document a pattern for that? Something like create your own `dsl` module somewhere as follows:

```rust
mod dsl {
    /// export all of diesel dsl
    pub use diesel::dsl::*;

    /// Export your extension types here
    pub use crate::your_extension::dsl::YourType;
}

```

File: diesel_derives/src/lib.rs
Comment (weiznich) [inline diesel_derives/src/lib.rs:1741]: I really like that pattern. That's really good work :heart:

File: diesel_derives/src/selectable.rs
Comment (weiznich) [inline diesel_derives/src/selectable.rs:151]: Can we use `quote_spanned!` here with a span of the original expression? That should result in the compiler pointing at the `#[diesel(select_expression = …)]` type in the relevant error message. (Also would be great to have a compile test for that)

**Review by weiznich — APPROVED** [review]
This looks great :tada:

I've left some minor stylistic comments otherwise I think the code is good. We should really add a changelog entry for this before merging.

File: diesel/src/query_source/aliasing/macros.rs
Comment (Ten0) [inline diesel/src/query_source/aliasing/macros.rs:89]: Took the option here to not put the implicit type alias in the explicit `const` syntax.
I'd like a second opinion on this. (I have no strong perference as I'm not using this syntax anyway.)

File: diesel_derives/src/lib.rs
Comment (Ten0) [reply diesel_derives/src/lib.rs:28]: I have some more of that in the dsl_auto_type crate:
https://github.com/Ten0/diesel/blob/71a5a4236e4dd860b1e0dd4ccfe42481b4b11743/dsl_auto_type/src/auto_type/mod.rs#L6-L13

Is that ok there or do you have a strong preference for changing it there as well?

File: diesel_derives/src/selectable.rs
Comment (Ten0) [reply diesel_derives/src/selectable.rs:151]: Actually quote_spanned only puts the span on the tokens that are directly written in the macro, but `#inferred_type` just keeps its span.
There was however an issue with the generated `_` placeholders though, which were not ideally spanned, so that is fixed in the type inferrence code.

File: diesel_derives/src/lib.rs
Comment (weiznich) [reply diesel_derives/src/lib.rs]: As long as it is consistent in a single crate I do not have strong preferences here.

File: diesel/src/query_source/aliasing/macros.rs
Comment (weiznich) [reply diesel/src/query_source/aliasing/macros.rs:89]: I think I would prefer to have it always there, just to have the always the same generated code. That makes it probably easier to work with that kind of code in the context of macros.

**Review by weiznich — APPROVED** [review]
:tada:

**Pull request merged**
